File: attached_assets/extracted/educational_analytics/backend/routes/analytics.py
from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required
import pandas as pd

from backend.ml import risk_model, features
from backend.decorators import roles_required

analytics_bp = Blueprint('analytics_bp', __name__)

@analytics_bp.route('/train', methods=['POST'])
@jwt_required()
@roles_required('admin')
def train_model():
    data = request.get_json()
    if not data or 'grades' not in data or 'attendance' not in data or 'labels' not in data:
        return jsonify({"msg": "Provide grades, attendance, and labels"}), 400

    grades_df = pd.DataFrame(data['grades'])
    attendance_df = pd.DataFrame(data['attendance'])
    labels_df = pd.DataFrame(data['labels'])  # expected columns: student_id, target (0/1)

    if 'student_id' not in labels_df.columns or 'target' not in labels_df.columns:
        return jsonify({"msg": "Labels must include student_id and target"}), 400

    features_df = features.build_features(grades_df, attendance_df)
    train_df = features_df.merge(labels_df, on='student_id', how='inner')

    if train_df.empty:
        return jsonify({"msg": "No matching students between features and labels"}), 400

    X = train_df[['performance_index']]
    y = train_df['target']

    risk_model.train(X, y)
    return jsonify({"msg": "Model trained and persisted"}), 200

@analytics_bp.route('/predict', methods=['POST'])
@jwt_required()
@roles_required('admin', 'lecturer', 'student')
def predict_risk():
    data = request.get_json()
    
    if 'grades' not in data or 'attendance' not in data:
        return jsonify({"msg": "Missing grades or attendance data"}), 400

    grades_df = pd.DataFrame(data['grades'])
    attendance_df = pd.DataFrame(data['attendance'])

    features_df = features.build_features(grades_df, attendance_df)

    if 'performance_index' not in features_df.columns:
        return jsonify({"msg": "Feature engineering failed"}), 500

    X = features_df[['performance_index']]

    try:
        predictions = risk_model.predict(X)
    except Exception:
        # Fallback: use rule-based probability from performance_index
        predictions = (1 - (X['performance_index'].clip(0, 100) / 100)).values

    features_df['risk_probability'] = predictions
    
    def map_risk(prob):
        if prob < 0.4:
            return 'LOW'
        elif prob < 0.7:
            return 'MEDIUM'
        else:
            return 'HIGH'

    features_df['risk_level'] = features_df['risk_probability'].apply(map_risk)

    # Align schema to dashboard expectations
    if 'percentage' in features_df.columns:
        features_df = features_df.rename(columns={'percentage': 'attendance'})
    if 'course' not in features_df.columns and 'course_id' in features_df.columns:
        features_df['course'] = features_df['course_id']

    return jsonify(features_df.to_dict(orient='records')), 200

@analytics_bp.route('/recommend', methods=['POST'])
@jwt_required()
@roles_required('admin', 'lecturer', 'student')
def recommend():
    data = request.get_json()
    if not data or 'students' not in data:
        return jsonify({"msg": "Provide students data"}), 400

    df = pd.DataFrame(data['students'])
    if 'risk_level' not in df.columns:
        return jsonify({"msg": "Each student must include risk_level"}), 400

    def recs(row):
        rl = row.get('risk_level', 'LOW')
        if rl == 'HIGH':
            return [
                'Schedule advisor meeting this week',
                'Attend all upcoming classes',
                'Join study group',
                'Book tutoring session'
            ]
        if rl == 'MEDIUM':
            return [
                'Review last 3 weeks material',
                'Increase weekly study hours',
                'Ask questions in lectures'
            ]
        return [
            'Explore advanced materials',
            'Mentor a peer',
            'Plan next semester courses'
        ]

    df['recommendations'] = df.apply(recs, axis=1)
    return jsonify(df[['student_id', 'course', 'risk_level', 'recommendations']].to_dict(orient='records')), 200

File: artifacts/api-server/src/routes/analytics.ts
import { Router, type IRouter } from "express";
import jwt from "jsonwebtoken";
import {
  PredictRiskBody,
  PredictRiskResponse,
  GetRecommendationsBody,
  GetRecommendationsResponse,
} from "@workspace/api-zod";

const router: IRouter = Router();

const JWT_SECRET = process.env.JWT_SECRET || "your-jwt-secret-change-in-production";

function requireAuth(req: any, res: any): { userId: number; role: string; email: string } | null {
  const authHeader = req.headers.authorization;
  if (!authHeader?.startsWith("Bearer ")) {
    res.status(401).json({ msg: "No token provided" });
    return null;
  }
  const token = authHeader.slice(7);
  try {
    return jwt.verify(token, JWT_SECRET) as { userId: number; role: string; email: string };
  } catch {
    res.status(401).json({ msg: "Invalid or expired token" });
    return null;
  }
}

function mapRisk(prob: number): "LOW" | "MEDIUM" | "HIGH" {
  if (prob < 0.4) return "LOW";
  if (prob < 0.7) return "MEDIUM";
  return "HIGH";
}

router.post("/predict", async (req, res) => {
  const user = requireAuth(req, res);
  if (!user) return;

  const parsed = PredictRiskBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ msg: "Missing grades or attendance data" });
    return;
  }

  const { grades, attendance } = parsed.data;

  const attendanceMap = new Map<number, number>();
  for (const a of attendance) {
    attendanceMap.set(a.student_id, a.percentage);
  }

  const predictions = grades.map((g) => {
    const att = attendanceMap.get(g.student_id) ?? 80;
    const performanceIndex = 0.7 * g.score + 0.3 * att;
    const riskProbability = Math.max(0, Math.min(1, 1 - performanceIndex / 100));
    return {
      student_id: g.student_id,
      course: g.course,
      score: g.score,
      attendance: att,
      risk_probability: Math.round(riskProbability * 1000) / 1000,
      risk_level: mapRisk(riskProbability),
    };
  });

  const data = PredictRiskResponse.parse(predictions);
  res.json(data);
});

router.post("/recommend", async (req, res) => {
  const user = requireAuth(req, res);
  if (!user) return;

  const parsed = GetRecommendationsBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ msg: "Provide students data" });
    return;
  }

  const { students } = parsed.data;

  const results = students.map((s) => {
    let recommendations: string[];
    if (s.risk_level === "HIGH") {
      recommendations = [
        "Schedule advisor meeting this week",
        "Attend all upcoming classes",
        "Join a study group",
        "Book a tutoring session",
      ];
    } else if (s.risk_level === "MEDIUM") {
      recommendations = [
        "Review last 3 weeks of material",
        "Increase weekly study hours",
        "Ask questions during lectures",
      ];
    } else {
      recommendations = [
        "Explore advanced materials",
        "Mentor a peer",
        "Plan next semester courses",
      ];
    }
    return {
      student_id: s.student_id,
      course: s.course,
      risk_level: s.risk_level,
      recommendations,
    };
  });

  const data = GetRecommendationsResponse.parse(results);
  res.json(data);
});

export default router;
